File: day02/main.lua
local lookup = {}

local function clear_table(t)
    for k in pairs(t) do
        t[k] = nil
    end
end

local function count_letter_occurences(id)
    local lk = lookup
    clear_table(lk)

    for i = 1, #id do
        local letter = id:sub(i, i)
        local count = lk[letter]
        if count == nil then
            lk[letter] = 1
        else
            lk[letter] = count + 1
        end
    end

    return lk
end

local function solve_part1(ids)
    local twos = 0
    local threes = 0
    for _, id in ipairs(ids) do
        local occurences = count_letter_occurences(id)
        local has2 = false
        local has3 = false
        for _, v in pairs(occurences) do
            has2 = has2 or v == 2
            has3 = has3 or v == 3
        end
        if has2 then twos = twos + 1 end
        if has3 then threes = threes + 1 end
    end
    return twos * threes
end

local function count_same_letters(a, b)
    -- assumes that a and b have the same length
    local count = 0
    local lastDifference = 0
    for i = 1, #a do
        if a:sub(i, i) == b:sub(i, i) then
            count = count + 1
        else
            lastDifference = i
        end
    end
    return count, lastDifference
end

local function solve_part2(ids)
    for i = 1, #ids - 1 do
        local a = ids[i]
        for j = i, #ids do
            local b = ids[j]
            local sameLetterCount, lastDifference = count_same_letters(a, b)
            local difference = #a - sameLetterCount
            if difference == 1 then
                return a:sub(1, lastDifference - 1) .. a:sub(lastDifference + 1)
            end
        end
    end
end

local function main()
    local file = io.open("day02/input.txt", "r")
    if not file then return end

    local ids = {}
    for line in file:lines() do
        table.insert(ids, line)
    end
    file:close()

    local part1 = solve_part1(ids)
    local part2 = solve_part2(ids)

    print(part1)
    print(part2)
end

main()

File: day02/main.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include <aoc/aoc.h>

typedef struct {
  char data[32];
} box_id;

#define AOC_T box_id
#define AOC_T_NAME BoxId
#include <aoc/array.h>

static void parse_line(char *line, size_t length, void *userData) {
  box_id id = {0};
  AocTrimRight(line, &length);
  AocMemCopy(id.data, line, length);
  AocArrayBoxIdPush(userData, id);
}

static int32_t solve_part1(const AocArrayBoxId *const ids) {
  int32_t twos = 0;
  int32_t threes = 0;
  const size_t idLength = strlen(ids->items[0].data);

  for (size_t i = 0; i < ids->length; ++i) {
    // count letter occurences
    int8_t occurences[26] = {0};
    for (size_t j = 0; j < idLength; ++j)
      occurences[ids->items[i].data[j] - 'a']++;

    // set flags for each occurence. Bit 2 means there is a character with 2
    // occurences and so on. can't be more than 32 because ids are smaller
    uint32_t counts = 0;
    for (size_t j = 0; j < 26; ++j)
      counts = AOC_SET_BIT(counts, occurences[j]);

    twos += AOC_CHECK_BIT(counts, 2);
    threes += AOC_CHECK_BIT(counts, 3);
  }
  return twos * threes;
}

static void solve_part2(const AocArrayBoxId *const ids, char output[const 32]) {
  const size_t idLength = strlen(ids->items[0].data);
  for (size_t i = 0; i < ids->length - 1; ++i) {
    const box_id *const a = &ids->items[i];
    for (size_t j = i + 1; j < ids->length; ++j) {
      const box_id *const b = &ids->items[j];
      int8_t differences = 0;
      size_t lastDifference = 0;
      for (size_t k = 0; k < idLength; ++k) {
        if (a->data[k] != b->data[k]) {
          differences++;
          lastDifference = k;
        }
      }
      if (differences == 1) {
        AocMemCopy(output, a->data, lastDifference);
        AocMemCopy(output + lastDifference, a->data + lastDifference + 1,
                   idLength - lastDifference);
        return;
      }
    }
  }
}

int main(void) {
  AocArrayBoxId ids = {0};
  AocArrayBoxIdCreate(&ids, 250);

  AocReadFileLineByLine("day02/input.txt", parse_line, &ids);

  const int32_t part1 = solve_part1(&ids);
  char part2[32] = {0};
  solve_part2(&ids, part2);

  printf("%d\n", part1);
  printf("%s\n", part2);

  AocArrayBoxIdDestroy(&ids);
}
